In [218]:
import warnings
warnings.filterwarnings('ignore')

In [219]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

In [220]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [221]:
!ls -l '/content/drive/My Drive/Colab Notebooks/datasets/Online_Retail.zip'

-rw------- 1 root root 22824989 Mar 11 00:24 '/content/drive/My Drive/Colab Notebooks/datasets/Online_Retail.zip'


In [222]:
!unzip /content/drive/My\ Drive/Colab\ Notebooks/datasets/Online_Retail.zip

Archive:  /content/drive/My Drive/Colab Notebooks/datasets/Online_Retail.zip
replace Online_Retail.xlsx? [y]es, [n]o, [A]ll, [N]one, [r]ename: ㅛ
error:  invalid response [ㅛ]
replace Online_Retail.xlsx? [y]es, [n]o, [A]ll, [N]one, [r]ename: ㅛ
error:  invalid response [ㅛ]
replace Online_Retail.xlsx? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: Online_Retail.xlsx      


In [223]:
!ls -l

total 23168
drwx------ 5 root root     4096 Mar 11 00:26 drive
-rw-r--r-- 1 root root 23715344 Mar  7 16:08 Online_Retail.xlsx
drwxr-xr-x 1 root root     4096 Mar  5 14:37 sample_data


In [224]:
%%time

DF = pd.read_excel('Online_Retail.xlsx')

DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB
CPU times: user 33.1 s, sys: 130 ms, total: 33.2 s
Wall time: 33.2 s


In [225]:
DF.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [226]:
DF = DF[DF['Quantity'] > 0]
DF = DF[DF['UnitPrice'] > 0]
DF = DF[DF['CustomerID'].notnull()]

DF.shape

(397884, 8)

In [227]:
DF.isnull().sum(axis=0)

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [228]:
DF['Country'].value_counts()[:10]

United Kingdom    354321
Germany             9040
France              8341
EIRE                7236
Spain               2484
Netherlands         2359
Belgium             2031
Switzerland         1841
Portugal            1462
Australia           1182
Name: Country, dtype: int64

In [229]:
DF = DF[DF['Country'] == 'United Kingdom']

DF.shape

(354321, 8)

## EDA

In [230]:
DF['StockCode'].value_counts()

85123A    1947
85099B    1464
22423     1426
84879     1333
47566     1307
          ... 
90214U       1
16161M       1
84016        1
90048        1
90038A       1
Name: StockCode, Length: 3645, dtype: int64

In [231]:
DF['CustomerID'].value_counts()

17841.0    7847
14096.0    5111
12748.0    4595
14606.0    2700
15311.0    2379
           ... 
14705.0       1
17923.0       1
15802.0       1
13747.0       1
15316.0       1
Name: CustomerID, Length: 3920, dtype: int64

In [232]:
DF['InvoiceNo'].value_counts()

576339    542
579196    533
580727    529
578270    442
573576    435
         ... 
539001      1
547792      1
575203      1
539005      1
555620      1
Name: InvoiceNo, Length: 16646, dtype: int64

In [233]:
DF['UnitPrice'].describe()

count    354321.000000
mean          2.963994
std          17.862655
min           0.001000
25%           1.250000
50%           1.950000
75%           3.750000
max        8142.750000
Name: UnitPrice, dtype: float64

In [234]:
DF.groupby('InvoiceNo')['Quantity', 'UnitPrice'].mean()

Quantity  UnitPrice
InvoiceNo                      
536365      5.714286   3.910000
536366      6.000000   1.850000
536367      6.916667   4.853333
536368      3.750000   4.775000
536369      3.000000   5.950000
...              ...        ...
581582      9.000000   1.765000
581583     38.000000   1.650000
581584     60.000000   1.285000
581585     13.238095   1.799048
581586     16.500000   5.057500

[16646 rows x 2 columns]

In [235]:
DF.groupby('CustomerID')['Quantity', 'UnitPrice'].mean()

Quantity  UnitPrice
CustomerID                         
12346.0     74215.000000   1.040000
12747.0        12.378641   4.367864
12748.0         5.603264   2.656279
12749.0         7.391960   4.999950
12820.0        12.237288   1.904746
...                  ...        ...
18280.0         4.500000   4.765000
18281.0         7.714286   5.622857
18282.0         8.583333   5.199167
18283.0         1.847884   1.614987
18287.0        22.657143   1.493571

[3920 rows x 2 columns]

In [236]:
DF1 = DF.groupby('CustomerID')['Quantity', 'UnitPrice'].sum()

In [237]:
DF1.head(10)

Quantity  UnitPrice
CustomerID                     
12346.0        74215       1.04
12747.0         1275     449.89
12748.0        25747   12205.60
12749.0         1471     994.99
12820.0          722     112.38
12821.0           70      14.99
12822.0          550     123.86
12823.0          230      38.25
12824.0          232      63.39
12826.0         1058     175.42

In [238]:
DF1.info()

<class 'pandas.core.frame.DataFrame'>
Float64Index: 3920 entries, 12346.0 to 18287.0
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Quantity   3920 non-null   int64  
 1   UnitPrice  3920 non-null   float64
dtypes: float64(1), int64(1)
memory usage: 91.9 KB


## K-means Clustering

In [239]:
from sklearn.cluster import KMeans

kmeans_1 = KMeans(n_clusters = 10,
                    init='k-means++',
                    max_iter=20,
                    random_state=2045)
kmeans_1.fit(DF1)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=20,
       n_clusters=10, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=2045, tol=0.0001, verbose=0)

In [240]:
kmeans_1.n_iter_

5

In [241]:
kmeans_1.cluster_centers_

array([[ 1589.61820652,   472.15265082],
       [60116.6       ,  1636.936     ],
       [19309.81818182,   678.83727273],
       [  313.8475672 ,   130.94521368],
       [ 4151.42613636,  1000.18738636],
       [16352.        , 33332.45      ],
       [32500.22222222,  2451.07555556],
       [75068.33333333,   383.11333333],
       [ 8891.15789474,   988.74736842],
       [24409.        , 16027.045     ]])

In [242]:
kmeans_1.labels_

array([7, 0, 9, ..., 3, 0, 0], dtype=int32)

In [243]:
kmeans_1.inertia_

1625390956.5589414

### Silhouette Analysis

In [244]:
DF1['Clustering'] = kmeans_1.labels_

In [245]:
DF1.head()

Quantity  UnitPrice  Clustering
CustomerID                                 
12346.0        74215       1.04           7
12747.0         1275     449.89           0
12748.0        25747   12205.60           9
12749.0         1471     994.99           0
12820.0          722     112.38           3

In [246]:
from sklearn.metrics import silhouette_samples

silhouette_samples(DF1, DF1['Clustering'])

array([0.6080873 , 0.4361024 , 0.36567924, ..., 0.83295802, 0.38348696,
       0.51154956])

In [247]:
DF1['Silh_Coef'] = silhouette_samples(DF1, DF1['Clustering'])

In [248]:
DF1.head()

Quantity  UnitPrice  Clustering  Silh_Coef
CustomerID                                            
12346.0        74215       1.04           7   0.608087
12747.0         1275     449.89           0   0.436102
12748.0        25747   12205.60           9   0.365679
12749.0         1471     994.99           0   0.463695
12820.0          722     112.38           3   0.541038

In [249]:
from sklearn.metrics import silhouette_score

silhouette_score(DF1, DF1['Clustering'])

0.6520786665221937

## Mean Shift

In [250]:
from sklearn.cluster import MeanShift

meanshift = MeanShift(bandwidth=900)
meanshift.fit_predict(DF1)

array([14,  0, 26, ...,  0,  0,  0])

In [251]:
from sklearn.cluster import estimate_bandwidth

estimate_bandwidth(DF1)

916.7749913604657

### Silhouette Analysis

In [252]:
DF1['Clustering'] = meanshift.labels_

In [253]:
DF1.head()

Quantity  UnitPrice  Clustering  Silh_Coef
CustomerID                                            
12346.0        74215       1.04          14   0.608087
12747.0         1275     449.89           0   0.436102
12748.0        25747   12205.60          26   0.365679
12749.0         1471     994.99           0   0.463695
12820.0          722     112.38           0   0.541038

In [254]:
from sklearn.metrics import silhouette_samples

silhouette_samples(DF1, DF1['Clustering'])

array([0.        , 0.54866652, 0.        , ..., 0.79379171, 0.08562067,
       0.45609166])

In [255]:
DF1['Silh_Coef'] = silhouette_samples(DF1, DF1['Clustering'])

In [256]:
DF1.head()

Quantity  UnitPrice  Clustering  Silh_Coef
CustomerID                                            
12346.0        74215       1.04          14   0.000000
12747.0         1275     449.89           0   0.548667
12748.0        25747   12205.60          26   0.000000
12749.0         1471     994.99           0   0.172422
12820.0          722     112.38           0   0.774385

In [257]:
from sklearn.metrics import silhouette_score

silhouette_score(DF1, DF1['Clustering'])

0.6601328209205919

## Result

* CustomerID로 groupby하고 Quantity, UnitPrice의 합계, k-means(n_clusters=10)

  0.6520786665222102

* CustomerID로 groupby하고 Quantity, UnitPrice의 합계, meanshift(bandwidth=900)

  0.6601328209205919